### PR TITLE
Add full violations codes to docs

### LINKED ISSUE #1145: Knowing the violation code it is quite hard to find the violation description
**Use Case.** I am testing wemake-python-styleguide on my project.

```
...
  22:5     WPS337 Found multiline conditions
  if (
  ^

Full list of violations and explanations:
https://wemake-python-stylegui.de/en/0.13.4/pages/usage/violations/
make: *** [infra/make/lint.mk:3: lint] Error 1
```

To get more information on what `WPS337` is for, I have to do several steps:

* Open https://wemake-python-stylegui.de/en/0.13.4/pages/usage/violations/
* Realize that WPS337 is probably behind the link titled `WPS300 - WPS399`
* And now I am stuck; I can `Ctrl + F` → `337`, but that won't help; so I try to find the word `multiline`
* and finally I reach the description I wanted: https://wemake-python-stylegui.de/en/0.13.4/pages/usage/violations/consistency.html#wemake_python_styleguide.violations.consistency.MultilineConditionsViolation

This takes time, especially when the title of the violation on this page does not exactly correspond to the name the tool prints on command line.

**Ideally,** I would see a link to the violation description right in the console, like this:

```
...
  22:5     WPS337 Found multiline conditions https://wps.it/0.13.4/WPS337/
  if (
  ^
...
```

Regrettably, wps.to domain is registered already, but wps.it is there.

**If that is not possible,** it would help if the docs pages at least included the violation codes, to make the page searchable via `Ctrl + F`.

### LINKED ISSUE #2288: [RFE] Include actual WPS code in docs
New to wemake here, great stuff.

I noticed that if I google WPS326 no good results are found, I suspect this is because in the docs, only '326' can be found.

I don't know if it's possible, but including a WPS326 somewhere might make it easier to go directly from the error I'm seeing to the docs where it is described if the WPSXXX code got picked up by a search engine.

It could go inline or even in the table up top.

Just a thought & suggestion.

Thanks for all this btw, my code is better now.

### PR DESCRIPTION
**Pull request by Bobronium**
# I made full violations codes indexable

Finally, people searching for `WPS527 Sovereign Starter Motor` will see that WPS truly means!
image

## Changes
- Full code is now prepended to `BaseViolation.__doc__`
- Full code is added to `BaseViolation.full_code` class var
- Summary is added to `BaseViolation.summary` class car
- `BaseViolation._full_name` is now a `classmethod`

## How it looks in docs
image
image

## Checklist
- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues
Closes #2288
Closes #1145

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: wemake_python_styleguide/violations/base.py [hunk 1]
```diff
@@ -67,6 +67,11 @@ class SomeViolation(ASTViolation):
 #: We use this type to define helper classes with callbacks to add violations.
 ErrorCallback = Callable[['BaseViolation'], None]
 
+def _prepend_skipping_whitespaces(prefix: str, text: str) -> str:
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:70]: nit: can you please move this to the very bottom of the file?

File: wemake_python_styleguide/violations/base.py [hunk 2]
```diff
@@ -67,6 +67,11 @@ class SomeViolation(ASTViolation):
...
+    lstripped_text = text.lstrip()
+    leading_whitespaces = text[:len(text) - len(lstripped_text)]
+    return f'{leading_whitespaces}{prefix}{lstripped_text}'
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:73]:
```suggestion
    return leading_whitespaces + prefix + lstripped_text
```

File: wemake_python_styleguide/violations/base.py [hunk 3]
```diff
@@ -98,12 +103,28 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
+    full_code: ClassVar[str]
+    summary: ClassVar[str]
 
     # We use this code to show base metrics and thresholds mostly:
     postfix_template: ClassVar[ViolationPostfixes] = (
         ViolationPostfixes.bigger_than
     )
 
+    def __init_subclass__(cls) -> None:
+        """
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:117]:
```suggestion
        """Sets additional values for subclasses."""
```

File: wemake_python_styleguide/violations/base.py [hunk 4]
```diff
@@ -98,12 +103,28 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
+        Derives and sets additional values for subclasses
+        """
+        if hasattr(cls, 'code'):
+            # this is mostly done for docs to display the full code,
+            # allowing its indexing in search engines and better discoverability
+            cls.full_code = cls._full_code()
+            cls.summary = cls.__doc__.lstrip().split('\n', maxsplit=1)[0]
+            # this hack adds full code to summary table in the docs
+            cls.__doc__ = _prepend_skipping_whitespaces(f'{cls.full_code} — ', cls.__doc__)
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:126]: We don't use `f` strings in this codebase 🙂

File: wemake_python_styleguide/violations/base.py [hunk 5]
```diff
@@ -98,12 +103,28 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
     # We use this code to show base metrics and thresholds mostly:
     postfix_template: ClassVar[ViolationPostfixes] = (
         ViolationPostfixes.bigger_than
     )
 
+    def __init_subclass__(cls) -> None:
+        """
+        Derives and sets additional values for subclasses
+        """
+        if hasattr(cls, 'code'):
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:120]:
```suggestion
        violation_code = getattr(cls, 'code', None)
        if violation_code is not None:
```

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thank you, this is amazing!

There are several nitpicks about code style, but other than that - I am happy to merge this! 🎉

File: wemake_python_styleguide/violations/base.py [hunk 6]
```diff
@@ -99,11 +98,28 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
+    # assigned in __init_subclass__
+    full_code: ClassVar[str]
+    summary: ClassVar[str]
+
     # We use this code to show base metrics and thresholds mostly:
     postfix_template: ClassVar[ViolationPostfixes] = (
         ViolationPostfixes.bigger_than
     )
 
+    def __init_subclass__(cls) -> None:
```
Comment (Bobronium) [inline wemake_python_styleguide/violations/base.py:110]: This causes `WPS338` error. Considering that #2411 is in place, what should be our course of action?

I see the following options:
1. Suppress the error until #2411 gets released
2. Obey current rules of ordering, refactor once #2411 gets released
3. Do nothing, wait until #2411 gets released

File: wemake_python_styleguide/violations/base.py [hunk 7]
```diff
@@ -99,11 +98,28 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
+        """Sets additional values for subclasses."""
+        violation_code = getattr(cls, 'code', None)
+        if violation_code is not None and cls.__doc__:
```
Comment (Bobronium) [inline wemake_python_styleguide/violations/base.py:113]: I'm going to rewrite this to early return now

File: wemake_python_styleguide/violations/base.py [hunk 8]
```diff
@@ -99,11 +99,34 @@ class BaseViolation(object, metaclass=abc.ABCMeta):
...
+    # assigned in __init_subclass__
+    full_code: ClassVar[str]
+    summary: ClassVar[str]
+
     # We use this code to show base metrics and thresholds mostly:
     postfix_template: ClassVar[ViolationPostfixes] = (
         ViolationPostfixes.bigger_than
     )
 
+    def __init_subclass__(cls, **kwargs) -> None:
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/base.py:111]:
```suggestion
    # TODO: remove `noqa` after a new release: 
    def __init_subclass__(cls, **kwargs) -> None:  # noqa: WPS338
```

**Review by sobolevn — APPROVED** [review]

**Comment by sobolevn** [comment]
Thank you!

**Pull request merged**